### PR TITLE
Pytorch port of SAC

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

Add both discrete and continuous SAC.

There's a bit of change in the policy/models to pass the action tensors directly, which allows the tanh caching to work (avoids numerical instability in inverse tanh). 

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

**Comment by vincentpierre** [comment]
I tried it with 3DBall and it worked. 
Do we want to integrate this change with `experiment_torch.py` before or after merge ?

**Comment by ervteng** [comment]
Let's do after. Currently I'm working on discrete, we can also integrate after that is done too

**Comment by ervteng** [comment]
Added discrete SAC as well.

File: ml-agents/mlagents/trainers/distributions_torch.py [hunk 1]
```diff
@@ -68,10 +101,13 @@ def __init__(self, hidden_size, num_outputs, conditional_sigma=False, **kwargs):
     def forward(self, inputs):
         mu = self.mu(inputs)
         if self.conditional_sigma:
-            log_sigma = self.log_sigma(inputs)
+            log_sigma = torch.clamp(self.log_sigma(inputs), min=-20, max=2)
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/distributions_torch.py:104]: Should these be consts?

File: ml-agents/mlagents/trainers/models_torch.py [hunk 2]
```diff
@@ -142,6 +179,116 @@ def forward(self, vec_inputs, vis_inputs, memories=None, sequence_length=1):
         return embedding, memories
 
 
+class QNetwork(NetworkBody):
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/models_torch.py:182]: The actor critic contains a NetworkBody, whereas this class inherits from one. Is there a reason for the difference? If there is a strong enough reason, maybe we should change ActorCritic to use the same pattern here?

Comment (awjuliani) [reply ml-agents/mlagents/trainers/models_torch.py:182]: Or alternatively use the AC pattern here.

File: ml-agents/mlagents/trainers/models_torch.py [hunk 3]
```diff
@@ -142,6 +179,116 @@ def forward(self, vec_inputs, vis_inputs, memories=None, sequence_length=1):
...
+        if self.use_lstm:
+            self.lstm = nn.LSTM(self.h_size, self.m_size // 2, 1)
+        if act_type == ActionType.DISCRETE:
+            self.q_heads = ValueHeads(
+                stream_names, network_settings.hidden_units, sum(act_size)
+            )
+        else:
+            self.q_heads = ValueHeads(stream_names, network_settings.hidden_units)
+
+    def forward(  # pylint: disable=W0221
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/models_torch.py:241]: Actually, it seems. that there is a lot of duplication of NetworkBody's forward function here.

**Comment by ervteng** [comment]
> Added some comments. Not sure how thorough we should be here, since on the one hand this is only for performance evaluation, and on the other there is a good change this will form the basis for whatever pytorch SAC looks like down the line.

I actually want to have a deeper discussion about this - I made this with minimal changes to the rest of the PyTorch code, but there are a couple things that I think we should do to the `models_torch.py` that would make it more modular. For instance, separating out the normalizers from the NetworkBody, and having ActorCritic be made up of two modules, an Actor and a Critic, so that we can make an Actor-only policy.

File: ml-agents/mlagents/trainers/models_torch.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/models_torch.py:241]: The main reason for this is the normalizers that are built into the NetworkBody - we don't want to normalize the action so I modified the forward to do this (this is also why I wasn't able to make this contain a NetworkBody). Ideally I'd like it to contain a NetworkBody and re-use it's forward, but to do that we need to first separate the Normalizers from the NetworkBody. 

So a class like Critic would have `Normalizers -> NetworkBody -> ValueHeads` and an Actor would have `Normalizers -> NetworkBody -> Distribution`. A Q-Network would then become `Normalizers + Action -> NetworkBody -> ValueHeads`.
